lnrpc/routerrpc/router_server.go

```diff
@@ -625,14 +630,27 @@ func (s *Server) probePaymentRequest(ctx context.Context, paymentRequest string,
 }
 
 // isLSP checks if the route hints indicate an LSP. An LSP is indicated with
-// true if the last node in each route hint has the same node id, false
-// otherwise.
-func isLSP(routeHints [][]zpay32.HopHint) bool {
+// true if the destination hop hint in each route hint has the same node id,
+// false otherwise. If the destination hop hint of the first route hint contains
+// a public channel, the function returns false because we can directly send a
```

Comment:
For my understanding, *if* there is a public channel in the route hints, it is *always* in the first route hint?

Reply:
Not necessarily, I think it would be more robust if we'd check if we know about *any* last hop to exclude an LSP, because a receiver could in principle mix private and public channels. I think it's also not too inefficient since we only have to check a few hints most of the time. So we should catch the case `[[private via nodeX], [public via nodeX], [public via nodeX]]` to not classify `nodeX` as an LSP, which could be added as a test.

Reply:
Thanks @gijswijs and @bitromortac for this improvement to the robustness of `isLSP`. I've added the public channel check for each final hop hint of each route hint as suggested and added a testcase to cover it.